auth_saml/models/res_users.py

```diff
@@ -0,0 +1,243 @@
+        """
+        token_osv = self.env['auth_saml.token']
+        saml_uid = validation['user_id']
+
+        user_ids = self.search(
+            [('saml_uid', '=', saml_uid), ('saml_provider_id', '=', provider)])
+
+        if not user_ids:
+            raise AccessDenied()
+
+        # TODO replace assert by proper raise... asserts do not execute in
+        # production code...
+        assert len(user_ids) == 1
+        user = user_ids[0]
+
+        # now find if a token for this user/provider already exists
+        token_ids = token_osv.search(
+            [('saml_provider_id', '=', provider), ('user_id', '=', user.id)])
+
+        if token_ids:
+            token_ids.write({'saml_access_token': saml_response})
+        else:
+            token_osv.create({'saml_access_token': saml_response,
+                              'saml_provider_id': provider,
+                              'user_id': user.id})
+
+        return user.login
+
+    @api.model
+    def auth_saml(self, provider, saml_response):
+
+        validation = self._auth_saml_validate(provider, saml_response)
+
+        # required check
+        if not validation.get('user_id'):
+            raise AccessDenied()
+
+        # retrieve and sign in user
+        login = self._auth_saml_signin(provider, validation, saml_response)
+
+        if not login:
+            raise AccessDenied()
+
+        # return user credentials
+        return self.env.cr.dbname, login, saml_response
+
+    def _check_credentials(self, token):
+        """Override to handle SAML auths.
+
+        The token can be a password if the user has used the normal form...
+        but we are more interested in the case when they are tokens
+        and the interesting code is inside the "except" clause.
+        """
+
+        try:
+            # Attempt a regular login (via other auth addons) first.
+            super(ResUser, self)._check_credentials(token)
+
+        except (AccessDenied, passlib.exc.PasswordSizeError):
+            # since normal auth did not succeed we now try to find if the user
+            # has an active token attached to his uid
+            res = self.env['auth_saml.token'].sudo().search(
+                [('user_id', '=', self.env.user.id),
+                 ('saml_access_token', '=', token)])
+
+            # if the user is not found we re-raise the AccessDenied
+            if not res:
+                # TODO: maybe raise a defined exception instead of the last
+                # exception that occurred in our execution frame
+                raise
+
+    @api.multi
+    def write(self, vals):
+        """Override to clear out the user's password when setting an SAML user
+        ID (as they can't cohabit).
+        """
+
+        # Clear out the pass when:
+        # - An SAML ID is being set.
+        # - The user is not the Odoo admin.
+        # - The "allow both" setting is disabled.
+        if (vals and vals.get('saml_uid') and self.id is not SUPERUSER_ID and
```

Comment:
What if `saml_uid` was already present in the record?

You should use a constraint instead.